modules/tinymce/src/core/main/ts/lookup/UserLookup.ts

```diff
@@ -0,0 +1,90 @@
+import { Obj } from '@ephox/katamari';
+
+import Editor from '../api/Editor';
+import * as Options from '../api/Options';
+
+/**
+ * TinyMCE User Lookup API
+ * Handles user information retrieval and caching.
+ *
+ * @class tinymce.UserLookup
+ * @example
+ * // Retrieve the current user ID
+ * tinymce.activeEditor.userLookup.getCurrentUserId();
+ *
+ * // Fetch user information by ID
+ * tinymce.activeEditor.userLookup.fetchUserById('user-id').then((user) => {
+ *  if (user) {
+ *   console.log('User found:', user);
+ *  };
+ * }).catch((error) => {
+ *  console.error('Error fetching user:', error);
+ * });
+ */
+
+type UserId = string;
+
+export interface User {
+  id: UserId;
+  name?: string;
+  avatar?: string;
+  description?: string;
+  [key: string]: any;
+}
+
+interface UserLookup {
+  /**
+   * Retrieves the current user ID from the editor.
+   *
+   * @method getCurrentUserId
+   * @return {string} The current user ID.
+   */
+  getCurrentUserId: () => UserId;
+
+  /**
+   * Fetches user information using a provided ID.
+   *
+   * @method fetchUserById
+   * @param {string} id - The user ID to fetch.
+   * @return {Promise<User>} A promise that resolves to the user information.
+   */
+  fetchUserById: (id: UserId) => Promise<User>;
+}
+
+const UserLookup = (editor: Editor): UserLookup => {
+  const userCache: Record<UserId, User> = {};
+
+  const lookup = (userId: UserId) =>
+    Obj.get(userCache, userId);
+
+  const store = (user: User, userId: UserId) => {
+    userCache[userId] = user;
+  };
+
+  const fetchUserById = (userId: UserId): Promise<User> =>
+    new Promise((resolve, reject) =>
+      lookup(userId)
+        .fold(() =>
+          Options.getFetchUserById(editor)(userId)
```

Comment:
This needs a lot more error handling. What if the user is not returning a promise in the function, what if the value in the promise doesn't have the expected properties. I would suggest at least doing some basic validations of the types I see that we already include boulder in the `autocompleter` in core so we could use that then we get pretty good error messages.

We have a `Type.isPromiseLike` so one could wrap that call in something like:
```
const fetchUserById = (editor: Editor, userId: UserId) {
  const fetchUserById = Options.getFetchUserById(editor);
  const result = fetchUserById(userId);
  if (Type.isPromiseLike(result)) {
    return result.then(schemaValidateWithBoulder});
  } else {
    return fail('fetch_user_by_id should return a promise';
  }
}
```